pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,835 @@
+)
+
+// Actuator acts upon ManagedSeed resources.
+type Actuator interface {
+	// CreateOrUpdate reconciles ManagedSeed creation or update.
+	CreateOrUpdate(context.Context, *seedmanagementv1alpha1.ManagedSeed, *gardencorev1beta1.Shoot) error
+	// Delete reconciles ManagedSeed deletion.
+	Delete(context.Context, *seedmanagementv1alpha1.ManagedSeed, *gardencorev1beta1.Shoot) error
+}
+
+type actuator struct {
+	gardenClient kubernetes.Interface
+	seedClient   kubernetes.Interface
+	shootClient  kubernetes.Interface
+	config       *config.GardenletConfiguration
+	imageVector  imagevector.ImageVector
+	logger       *logrus.Logger
+}
+
+func newActuator(gardenClient, seedClient, shootClient kubernetes.Interface, config *config.GardenletConfiguration, imageVector imagevector.ImageVector, logger *logrus.Logger) Actuator {
+	return &actuator{
+		gardenClient: gardenClient,
+		seedClient:   seedClient,
+		shootClient:  shootClient,
+		config:       config,
+		imageVector:  imageVector,
+		logger:       logger,
+	}
+}
+
+// CreateOrUpdate reconciles ManagedSeed creation or update.
+func (a *actuator) CreateOrUpdate(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	managedSeedLogger := logger.NewFieldLogger(a.logger, "managedSeed", kutil.ObjectName(managedSeed))
+
+	// Create garden namespace in the shoot
+	managedSeedLogger.Infof("Creating garden namespace in shoot %s", kutil.ObjectName(shoot))
+	gardenNamespace := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.GardenNamespace}}
+	if err := a.shootClient.Client().Create(ctx, gardenNamespace); err != nil && !apierrors.IsAlreadyExists(err) {
+		return fmt.Errorf("could not create garden namespace in shoot %q: %+v", kutil.ObjectName(shoot), err)
+	}
+
+	switch {
+	case managedSeed.Spec.Seed != nil:
+		// Register the shoot as seed
+		managedSeedLogger.Infof("Registering shoot %s as seed", kutil.ObjectName(shoot))
+		if err := a.registerAsSeed(ctx, managedSeed, shoot); err != nil {
+			return fmt.Errorf("could not register shoot %q as seed: %+v", kutil.ObjectName(shoot), err)
+		}
+	case managedSeed.Spec.Gardenlet != nil:
+		// Deploy gardenlet into the shoot, it will register the seed automatically
+		managedSeedLogger.Infof("Deploying gardenlet into shoot %s", kutil.ObjectName(shoot))
+		if err := a.deployGardenlet(ctx, managedSeed, shoot); err != nil {
+			return fmt.Errorf("could not deploy gardenlet into shoot %q: %+v", kutil.ObjectName(shoot), err)
+		}
+	}
+
+	// Set managed seed reference in the shoot
+	if shoot.Spec.ManagedSeedName == nil {
+		managedSeedLogger.Infof("Setting managed seed reference in shoot %s", kutil.ObjectName(shoot))
+		shoot.Spec.ManagedSeedName = &managedSeed.Name
+		if err := a.gardenClient.Client().Update(ctx, shoot); err != nil {
+			return fmt.Errorf("could not set managed seed reference in shoot %q: %+v", kutil.ObjectName(shoot), err)
+		}
+	}
```

Comment:
Should we already set this before the switch?

Reply:
We can do it at any time during the managed seed reconciliation, so before or after the "main" part, the important detail is that doing it triggers a reconcile of the shoot (we actually rely on this for the `APIServer` settings to be applied). Do we want to trigger this reconciliation of the shoot rather at the beginning or at the end of a managed seed reconciliation? I thought the beginning was better, because then the 2 reconciliations go in parallel and the desired state is reached earlier.

Reply:
> (we actually rely on this for the APIServer settings to be applied).

This only works for the very first registration, right? What if the API server settings are changed later again? Then no reconciliation will be triggered.

Reply:
Yes, you are right, and also for this reason I agree with @timebertt that this mechanism should be reworked entirely, see https://github.com/gardener/gardener/pull/3177#discussion_r544344146.